sympy/matrices/dense.py

```diff
@@ -184,13 +184,16 @@ def _eval_matrix_mul(self, other):
             # cache self._mat and other._mat for performance
             mat = self._mat
             other_mat = other._mat
+            poly = False
+            for i in range(len(mat)):
+                poly = poly | mat[i].is_Poly
```

Comment:
Don't use bitwise operators for program logic. Also usually `for i in range(len(X)): <stuff with X[i]>` can be replaced with `for i in X: <stuff with i>`. 

Actually this whole thing can be simplified to just `poly = any(i.is_Poly for i in mat)`. Or maybe you could just check the first element (is it possible to have mismatched types in a matrix?).

Reply:
I explain why I check all in my comment.

Reply:
Also, the reason why I didnt use any is because I thought there will be two loops over the elements. I'm just checking this.

Reply:
"is it possible to have mismatched types in a matrix?": Yes the unit test failed when only checking the first element.

Reply:
> the unit test failed when only checking the first element.

Which test was that?
I think that users of polynomial matrices should take care of turning all entries into polynomials.